### PR TITLE
Generic SSL Interface

### PR DESCRIPTION
**Pull request by jimktrains**
This allows for the SSL Context to be created by the end developer. (Built to allow custom SNI handling.)

### CODE REVIEW

File: src/iron.rs [hunk 1]
```diff
@@ -6,9 +6,15 @@ use std::time::Duration;
...
 
+#[cfg(feature = "ssl")]
+use hyper::net::{NetworkStream, Openssl, Ssl};
+
 pub use hyper::server::Listening;
 use hyper::server::Server;
-use hyper::net::Fresh;
+use hyper::net::{Fresh, NetworkListener, HttpListener};
+
+#[cfg(feature = "ssl")]
```
Comment (sfackler) [inline src/iron.rs]: The ssl feature doesn't seem super necessary.

File: src/iron.rs [hunk 2]
```diff
@@ -87,11 +90,68 @@ impl Protocol {
...
+    fn protocol(&self) -> Protocol { Protocol::Http }
+
+    fn create_server(&self, sock_addr: SocketAddr) -> Result<Server, Error>{
+        Server::http(sock_addr)
+   }
+}
+
+/// Default HTTPs handler
+#[cfg(feature = "ssl")]
+pub struct HttpsProtocolHandler<S: Ssl + Clone + Send> {
```
Comment (sfackler) [inline src/iron.rs]: I'd just give this a constructor that takes in the `S` and avoid the Option + `impl HttpsProtocolHandler<Openssl>` altogether.

**Comment by jimktrains** [comment]
Travis is failing because I didn't commit a dependency to openssl in Cargo.toml, but use it in the code. That may need to be refactored out, or the HTTPS server moved somewhere else?

File: src/iron.rs (hunk 1 above)
Comment (jimktrains) [inline src/iron.rs]: This is a hold over from the current codebase; I didn't add the feature.

**Comment by jimktrains** [comment]
Also, I'll eventually squash all of this, or rewrite some of the commit messages. I just realized my first one is non-informative.

File: src/iron.rs [hunk 3]
```diff
@@ -87,11 +90,68 @@ impl Protocol {
...
-            Protocol::Https { .. } => "https"
+            Protocol::Https => "https"
         }
     }
 }
 
+/// Protocol Handlers create servers
+pub trait ProtocolHandler<TListener: NetworkListener> {
+    /// Return the name used for this protocol in a URI's scheme part.
+    fn name(&self) -> &'static str;
```
Comment (samgiles) [inline src/iron.rs]: Do we need this if `protocol().name()` returns the same thing?

**Comment by posix4e** [comment]
Do we still want this?

**Comment by sfackler** [comment]
Yes!

**Comment by reem** [comment]
Closing because stale, but please feel free to make a new PR with updated code!

**Comment by jimktrains** [comment]
I was coming back to this, but it looks like there is now a way to pass in a TLS server to handle the TLS-bits.

**Pull request closed without merging**